### PR TITLE
Inline duration building in shift/2 functions

### PR DESCRIPTION
**Pull request by sabiwara**
Close https://github.com/elixir-lang/elixir/issues/13475

### CODE REVIEW

**Review by whatyouhide — APPROVED** [review]
Sweet 🍬

File: lib/elixir/lib/calendar/time.ex [hunk 1]
```diff
@@ -592,11 +592,13 @@ defmodule Time do
     }
   end
 
-  defp new_duration!(%Duration{} = duration) do
+  @doc false
+  def new_duration!(%Duration{} = duration) do
```
Comment (josevalim) [inline lib/elixir/lib/calendar/time.ex:596]: We should use `__` so they are not accidentally imported. Maybe `__duration__!`?

File: lib/elixir/src/elixir_erl_pass.erl [hunk 2]
```diff
@@ -631,6 +636,30 @@ rewrite_strategy(erlang, Right, Args) ->
...
+  Left == 'Elixir.NaiveDateTime';
+  Left == 'Elixir.Time'
+->
+  case basic_type_arg(Opts) of
+    true ->
+      {Module, Fun} =
+        case (Left == 'Elixir.Date' orelse Left == 'Elixir.Time') of
+          % Time and Date have more restrictive validations than Elixir.Duration
+          true -> {Left, 'new_duration!'};
+          false -> {'Elixir.Duration', 'new!'}
```
Comment (josevalim) [inline lib/elixir/src/elixir_erl_pass.erl:651]: One option is to define a `__duration__!` on all of them. Your call. :)

**Review by josevalim — APPROVED** [review]
One tiny comment and ship it :)

File: lib/elixir/src/elixir_erl_pass.erl (hunk 2 above)
Comment (sabiwara) [reply lib/elixir/src/elixir_erl_pass.erl:651]: I hesitated doing so but it feels cleaner indeed. Will do it!

**Comment by sabiwara** [comment]
I also confirmed we weren't being paranoid and [benched](https://github.com/sabiwara/elixir_benches/commit/d56b986cb910796d0d5176ff4db59a1e68455abe) the benefit justifies it (ran on main without the optimization oc).
Looks it's worth it!

File: lib/elixir/lib/calendar/time.ex (hunk 1 above)
Comment (sabiwara) [reply lib/elixir/lib/calendar/time.ex]: Done in [40e8d10](https://github.com/elixir-lang/elixir/pull/13489/commits/40e8d106594bc5bd75dd56616a9fa322aae4b4e6)

File: lib/elixir/src/elixir_erl_pass.erl (hunk 2 above)
Comment (sabiwara) [reply lib/elixir/src/elixir_erl_pass.erl]: Done in [40e8d10](https://github.com/elixir-lang/elixir/pull/13489/commits/40e8d106594bc5bd75dd56616a9fa322aae4b4e6)

**Review by josevalim — APPROVED** [review]

**Pull request merged**